lib/fibonacci.rb

```diff
@@ -1,9 +1,36 @@
 # Improved Fibonacci
 
-# Time Complexity - ?
-# Space Complexity - ?  (should be O(n))
+# Time Complexity - O(n)
+# Space Complexity - O(n)
 # Hint, you may want a recursive helper method
 # Hint:  Fib(0) = 0, Fib(1) = 1, work up from there
-def fibonacci(n)
-  raise NotImplementedError, "Please implement this method for the project"
+# Calculates each fib(n) only once, stores the answers we already know
+#
+def fibonacci(n, hash = {})
```

Comment:
This works, but do you need to maintain a hash with roughly n key-value pairs?  Can you do it with just maintaining two numbers?